### DISCUSSION TITLE
Ability to specify configuration file as command-line argument for unified-language-server

### DISCUSSION THREAD

**Discussion by hjpotter92** [unifiedjs/unified#133] [Q&A] [answered]
Copying over issue from the project: https://github.com/unifiedjs/unified-language-server/issues/8#issue-859514136

### Subject of the feature

Better documentation around handling of configuration file, or specifying config file as an argument when starting server

### Problem

Currently, the docs do not make it clear where and how the configuration file is supposed to be sent to server.

**Comment by ChristianMurphy** [comment]
https://github.com/unifiedjs/unified-language-server/blob/main/CONFIGURATION.md#for-neovim

Documents where to add the configuration file.
settings.json in the home folder

**Reply by hjpotter92** [reply]
> https://github.com/unifiedjs/unified-language-server/blob/main/CONFIGURATION.md#for-neovim
> 
> Documents where to add the configuration file.
> settings.json in the home folder

How would this translate when I'm creating a new client for emacs?

**Reply by ChristianMurphy** [reply]
The default settings, and custom settings are merged here: https://github.com/unifiedjs/unified-language-server/blob/ca717b1322a9ff09cbea3442505929a7ae00327c/src/server.js#L153-L157

> How would this translate when I'm creating a new client for emacs?

However you want it to, the server just needs a configuration with a settings object, how a client chooses to allow the settings to be set/changed is flexible.

**Reply by hjpotter92** [reply]
> However you want it to, the server just needs a configuration with a settings object, how a client chooses to allow the settings to be set/changed is flexible.

This brings me back to my original question. emacs-lsp uses the stdio pipe processing for providing the LSP support. The command used to start the server is `unified-language-server --parser=remark-parse --stdio`.

There is no 'default' settings file name, no argument to specify one, and just a rough declaration of a json.

I come from python's environment, where for nearly each server implementation, the config is read from `pyproject.toml` or `pyrightconfig.json` etc. (I think something similar happens with ts-ls server etc.)

**Reply by ChristianMurphy** [reply]
> the config is read from

are you asking for what the community convention is?

remark uses `.remarkrc` (with optional `js` or `json`) extensions or a `remark` filed in package.json, in the project root.
rehype uses `.rehyperc` (with optional `js` or `json`) extensions or a `remark` filed in package.json, in the project root.

@aecepoglu may be able to comment more on the mix of both config described in the example vim file.

**Reply by aecepoglu** [reply]
I tried to have a look at emacs-lsp's settings page but I kept getting a 404, so I cannot comment too confidently.
But yes, we rely on the client passing configuration data as per Language Server Protocol specification https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_didChangeConfiguration

Within the data it receives, the server expects to see the shape `{settings: {unified-language-server: {...}}, ...}` 

As for your question: "how can I do this for emacs" I had a look a random item (lsp-rust) from this list https://emacs-lsp.github.io/lsp-mode/page/languages/ . I don't know how emacs works but it _seems_ for each language-server it also requires a middleware like so https://emacs-lsp.github.io/lsp-mode/page/lsp-rust/

the language-server still expects configuration the same way we do: https://github.com/rust-lang/rls
`lsp-rust` reads config from wherever and passes it to `rls` as the LSP requires.

This sounds like it requires way more effort than language-server integrations are for other text-editors / IDEs but I suppose that's what you'll have to do.

**Reply by aecepoglu** [reply]
And I also had a look at python because you mentioned it.

https://github.com/palantir/python-language-server#configuration

It mentions

> Overall configuration is computed first from user configuration (in home directory), overridden by configuration passed in by the language client, and then overriden by configuration discovered in the workspace.

For us it is simply 

> ... configuration passed in by the language client ...

**Comment by hjpotter92** [comment, marked as answer]
Thanks you @aecepoglu and @ChristianMurphy 

With your help, I've added a rudimentry client to emacs as well. https://github.com/emacs-lsp/lsp-mode/pull/2800
